pkg/runtime/scheduler/scheduler.go

```diff
@@ -0,0 +1,287 @@
+/*
+Copyright 2024 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"strings"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	schedulerv1pb "github.com/dapr/dapr/pkg/proto/scheduler/v1"
+	"github.com/dapr/dapr/pkg/scheduler"
+	schedulerclient "github.com/dapr/dapr/pkg/scheduler/client"
+	"github.com/dapr/dapr/pkg/security"
+	"github.com/dapr/kit/concurrency"
+	"github.com/dapr/kit/logger"
+)
+
+var log = logger.NewLogger("dapr.runtime.scheduler")
+
+// Client represents a client for interacting with the scheduler.
+type schedulerClient struct {
+	conn      *grpc.ClientConn
+	scheduler schedulerv1pb.SchedulerClient
+	address   string
+	security  security.Handler
+}
+
+// Manager manages connections to multiple schedulers.
+type Manager struct {
+	clients     []*schedulerClient
+	connDetails scheduler.SidecarConnDetails
+	lock        sync.Mutex
+	lastUsedIdx int
+}
+
+func NewManager(ctx context.Context, ns string, appID string, addresses string, sec security.Handler) *Manager {
+	sidecarDetails := scheduler.SidecarConnDetails{
+		Namespace: ns,
+		AppID:     appID,
+	}
+
+	manager := &Manager{
+		connDetails: sidecarDetails,
+	}
+
+	var schedulerHostPorts []string
+	if strings.Contains(addresses, ",") {
+		parts := strings.Split(addresses, ",")
+		for _, part := range parts {
+			hostPort := strings.TrimSpace(part)
+			schedulerHostPorts = append(schedulerHostPorts, hostPort)
+		}
+	} else {
+		schedulerHostPorts = []string{addresses}
+	}
+
+	for _, address := range schedulerHostPorts {
+		log.Debug("Attempting to connect to Scheduler")
+		conn, client, err := schedulerclient.New(ctx, address, sec)
+		if err != nil {
+			log.Debugf("Scheduler client not initialized for address: %s", address)
```

Comment:
I didn't return the error bc there is a check if the client and conn aren't good in the scheduler manager logic it retries establishing the conn. I didn't want the sidecar to error out in this case since it is retried indefinitely bc if the scheduler addresses are supplied it _should_ eventually establish the conn.